chasm/lib/scheduler/invoker_tasks.go

```diff
@@ -0,0 +1,603 @@
+)
+
+type (
+	InvokerTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		BaseLogger     log.Logger
+		HistoryClient  resource.HistoryClient
+		FrontendClient workflowservice.WorkflowServiceClient
+	}
+
+	InvokerExecuteTaskExecutor struct {
+		InvokerTaskExecutorOptions
+	}
+
+	InvokerProcessBufferTaskExecutor struct {
+		InvokerTaskExecutorOptions
+	}
+
+	// Per-task context.
+	invokerTaskExecutorContext struct {
+		context.Context
+
+		actionsTaken int
+		maxActions   int
+	}
+
+	rateLimitedError struct {
+		// The requested interval to delay processing by rescheduilng.
+		delay time.Duration
+	}
+)
+
+const (
+	// Lower bound for the deadline in which buffered actions are dropped.
+	startWorkflowMinDeadline = 5 * time.Second
+
+	// Because the catchup window doesn't apply to a manual start, pick a custom
+	// execution deadline before timing out a start.
+	manualStartExecutionDeadline = 1 * time.Hour
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	InvokerMaxStartAttempts = 10 // TODO - dial this up/remove it
+)
+
+var (
+	errRetryLimitExceeded       = errors.New("retry limit exceeded")
+	_                     error = &rateLimitedError{}
+)
+
+func NewInvokerExecuteTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerExecuteTaskExecutor {
+	return &InvokerExecuteTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func NewInvokerProcessBufferTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerProcessBufferTaskExecutor {
+	return &InvokerProcessBufferTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func (e *InvokerExecuteTaskExecutor) Validate(
+	_ chasm.Context,
+	invoker *Invoker,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) (bool, error) {
+	// If another execute task already happened to kick everything off, we don't need
+	// this one.
+	eligibleStarts := invoker.getEligibleBufferedStarts()
+	valid := len(invoker.GetTerminateWorkflows())+
+		len(invoker.GetCancelWorkflows())+
+		len(eligibleStarts) > 0
+	return valid, nil
+}
+
+func (e *InvokerExecuteTaskExecutor) Execute(
+	ctx context.Context,
+	invokerRef chasm.ComponentRef,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) error {
+	var result executeResult
+
+	var invoker *Invoker
+	var scheduler *Scheduler
+
+	// Read and deep copy returned components, since we'll continue to access them
+	// outside of this function (outside of the MS lock).
+	_, err := chasm.ReadComponent(
+		ctx,
+		invokerRef,
+		func(i *Invoker, ctx chasm.Context, _ any) (struct{}, error) {
+			invoker = &Invoker{
+				InvokerInternal: common.CloneProto(i.InvokerInternal),
+			}
+
+			s, err := i.Scheduler.Get(ctx)
+			if err != nil {
+				return struct{}{}, err
+			}
+			scheduler = &Scheduler{
+				SchedulerInternal:  common.CloneProto(s.SchedulerInternal),
+				cacheConflictToken: s.cacheConflictToken,
+				compiledSpec:       s.compiledSpec,
```

Comment:
Can you confirm and add a comment saying that the compiled spec is never mutated, only replaced.